lib/_http_incoming.js

```diff
@@ -160,19 +160,20 @@ IncomingMessage.prototype._read = function _read(n) {
     readStart(this.socket);
 };
 
-
 // It's possible that the socket will be destroyed, and removed from
 // any messages, before ever calling this.  In that case, just skip
 // it, since something else is destroying this connection anyway.
-IncomingMessage.prototype.destroy = function destroy(error) {
-  // TODO(ronag): Implement in terms of _destroy
-  this.destroyed = true;
-  if (this.socket)
-    this.socket.destroy(error);
-  return this;
+IncomingMessage.prototype._destroy = function _destroy(err, cb) {
+  if (!this.readableEnded || !this.complete) {
+    this.aborted = true;
+    this.emit('aborted');
+  }
+  if (this.socket && !this.readableEnded) {
+    this.socket.destroy(err);
+  }
+  this.listenerCount('error') > 0 ? cb(err) : cb();
 };
 
```

Comment:
I would probably do `_destroy` like this:

```js
if (!this.readableEnded || !this.complete) {
  this.aborted = true;
  this.emit('aborted');
  err = connResetException('aborted');
}

const callback = (er) => {
  cb(this.aborted && this.listenerCount('error') === 0 ? null : (er || err));
};

if (err && this.socket) {
  this.socket.destroy(err);
  finished(this.socket, callback);
} else {
  callback(err);
}

```

Reply:
`finished` would be https://nodejs.org/api/stream.html#stream_stream_finished_stream_options_callback?

I understand your point. Why do we have to wait for the socket to finish before calling `cb`? Is that related to the timing issue you talked about previously? I am just asking because I don't know.

Reply:
> understand your point. Why do we have to wait for the socket to finish before calling cb? Is that related to the timing issue you talked about previously? I am just asking because I don't know.

No, it's not necessary per se. Just best practice. You can also skip waiting it is just less "correct".

Reply:
Ok, thank you so much for the clarification

Reply:
Unfortunately, that implementation is causing some problems in tests. The pain points are 2:

1. tests that call `destroy()` without an error now have an `error` event emitted. For example, `test/parallel/test-async-wrap-tlssocket-asyncreset.js`

![Screenshot from 2020-11-19 07-08-20](https://user-images.githubusercontent.com/15520377/99628653-ded47c00-2a36-11eb-8489-d999e69cdcf0.png)

others are erroring with a timeout:

![Screenshot from 2020-11-19 07-17-11](https://user-images.githubusercontent.com/15520377/99628847-47235d80-2a37-11eb-8726-77eb8a83f2f7.png)

2. for some reason I am trying to figure out, it seems the `Stream.finished` callback is not getting called in the client in the test `parallel/test-http-client-spurious-aborted.js` causing the error not to be propagated:

![Screenshot from 2020-11-19 06-34-18](https://user-images.githubusercontent.com/15520377/99629147-c6b12c80-2a37-11eb-8890-3bd287eec558.png)

This implementation has just the problem `2`:

```js
IncomingMessage.prototype._destroy = function _destroy(err, cb) {
  if (!this.readableEnded || !this.complete) {
    this.aborted = true;
    this.emit('aborted');
  }
  if (this.socket && this.aborted) {
    this.socket.destroy(err);
    const cleanup = Stream.finished(this.socket, e => {
      cleanup()
      this.listenerCount('error') > 0 ? cb(e || err) : cb();
    })
  } else {
    this.listenerCount('error') > 0 ? cb(err) : cb();
  }
};
```
Please overlook the code repetition for now. I'll fix it once I find out the problem.

Reply:
I am debugging `test/parallel/test-http-client-spurious-aborted.js`.

It looks like [this check](https://github.com/nodejs/node/blob/master/lib/internal/streams/end-of-stream.js#L163) is always `false` on the client when running the test. On the server-side, that check returns `true` and `callback` is called on the next tick, while on the client-side, that doesn't happen.

Reply:
I tried to implement it as close as possible with the caveats explained earlier. I am now destroying the socket only if it hasn't already been destroyed. That avoids calling finished on a finished stream.

ptal @ronag